pyomo/contrib/mindtpy/single_tree.py

```diff
@@ -493,6 +500,48 @@ def __call__(self):
         self.handle_lazy_master_feasible_solution(
             master_mip, solve_data, config, opt)
 
+        # regularization is activated after the first feasible solution is found.
+        if config.add_regularization is not None and solve_data.best_solution_found is not None:
+            # the master problem might be unbounded, regularization is activated only when a valid bound is provided.
+            if config.add_cuts_at_incumbent:
+                self.copy_lazy_var_list_values(opt,
+                                            master_mip.MindtPy_utils.variable_list,
+                                            solve_data.mip.MindtPy_utils.variable_list,
+                                            config)
+                if config.strategy == 'OA':
+                    self.add_lazy_oa_cuts(
+                        solve_data.mip, None, solve_data, config, opt)
+
+            if (solve_data.objective_sense == minimize and solve_data.LB != float('-inf')) or (solve_data.objective_sense == maximize and solve_data.UB != float('inf')):
+                master_mip, master_mip_results = solve_master(
+                    solve_data, config, regularization_problem=True)
+                if master_mip_results.solver.termination_condition in {tc.optimal, tc.feasible}:
+                    handle_master_optimal(
+                        master_mip, solve_data, config, update_bound=False)
+                elif master_mip_results.solver.termination_condition is tc.infeasible:
+                    config.logger.info('Projection problem infeasible.')
+                    if config.reduce_level_coef:
+                        config.level_coef = config.level_coef / 2
+                        master_mip, master_mip_results = solve_master(solve_data, config, regularization_problem=True)
+                        config.level_coef = config.level_coef * 2
```

Comment:
I'm confused here. Do you restore the coefficient after solving the regularization problem in case that it is infeasible? This is done for a single iteration, but I would expect that if the regularization problem fails once with a coefficient, it will fail in the subsequent iterations. In fact, it shouldn't fail in the first place since we are initializing the regularization problem with the master solution. At the very least let's use the "reduction coefficient" as a parameter (here it is 2) so that anyone can play with it. This is one point we should mainly discuss (together with the fake bounds)